test/correctness/pytorch.cpp

```diff
@@ -0,0 +1,232 @@
+
+        std::cerr
+            << "Correct source code:\n"
+            << correct_src
+            << "Actual source code:\n"
+            << src
+            << "Difference starts at:" << diff << "\n"
+            << "Correct: " << correct_src.substr(diff, diff_end - diff) << "\n"
+            << "Actual: " << src.substr(diff, diff_end - diff) << "\n";
+
+        exit(1);
+    }
+}
+
+}  // namespace
+
+int main(int argc, char **argv) {
+    Param<float> alpha("alpha");
+    Param<int32_t> beta("beta");
+    Var x("x");
+
+    Func buf("buf");
+    buf(x) = cast<int32_t>(alpha + cast<float>(beta));
+
+    {
+        // TODO(mgharbi): test that Target("host-cuda") raises an exception since
+        // we require the "user_context" feature when using CUDA
+
+        std::string pytorch_out = Internal::get_test_tmp_dir() + "pytorch_test1.pytorch.h";
+        Internal::ensure_no_file_exists(pytorch_out);
+
+        std::vector<Argument> args{alpha, beta};
+        buf.compile_to({{Output::pytorch_wrapper, pytorch_out}}, args, "test1", Target("host"));
+
+        Internal::assert_file_exists(pytorch_out);
+        std::string actual = read_entire_file(pytorch_out);
+
+        std::string expected =
+            R"GOLDEN_CODE(#include "HalideBuffer.h"
+#include "HalidePyTorchHelpers.h"
+
+struct halide_buffer_t;
+struct halide_filter_metadata_t;
+
+#ifndef HALIDE_MUST_USE_RESULT
+#ifdef __has_attribute
+#if __has_attribute(nodiscard)
+#define HALIDE_MUST_USE_RESULT [[nodiscard]]
+#elif __has_attribute(warn_unused_result)
+#define HALIDE_MUST_USE_RESULT __attribute__((warn_unused_result))
+#else
+#define HALIDE_MUST_USE_RESULT
+#endif
+#else
+#define HALIDE_MUST_USE_RESULT
+#endif
+#endif
+
+#ifndef HALIDE_FUNCTION_ATTRS
+#define HALIDE_FUNCTION_ATTRS
+#endif
+
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+HALIDE_FUNCTION_ATTRS
+int test1(float _alpha, int32_t _beta, struct halide_buffer_t *_buf_buffer);
+
+HALIDE_FUNCTION_ATTRS
+int test1_argv(void **args);
+
+HALIDE_FUNCTION_ATTRS
+const struct halide_filter_metadata_t *test1_metadata();
+
+#ifdef __cplusplus
+}  // extern "C"
+#endif
+
+HALIDE_FUNCTION_ATTRS
+inline int test1_th_(float _alpha, int32_t _beta, at::Tensor &_buf) {
+    // Check tensors have contiguous memory and are on the correct device
+    HLPT_CHECK_CONTIGUOUS(_buf);
+
+    // Wrap tensors in Halide buffers
+    Halide::Runtime::Buffer<int32_t> _buf_buffer = Halide::PyTorch::wrap<int32_t>(_buf);
+
+    // Run Halide pipeline
+    int err = test1(_alpha, _beta, _buf_buffer);
+
+    AT_ASSERTM(err == 0, "Halide call failed");
+    return 0;
+}
+)GOLDEN_CODE";
+
+        compare_src(actual, expected);
+    }
+
+    {
+        std::string pytorch_out = Internal::get_test_tmp_dir() + "pytorch_test2.pytorch.h";
+        Internal::ensure_no_file_exists(pytorch_out);
+
+        std::vector<Argument> args{alpha, beta};
+        buf.compile_to({{Output::pytorch_wrapper, pytorch_out}}, args, "test2", Target("host-cuda-user_context"));
```

Comment:
What is "host-cuda" going to do on systems that don't support cuda? I guess it just won't turn on any flags after failing to find the cuda drivers.

Can we just use something explicit like linux-x86-64-cuda?

Reply:
> What is "host-cuda" going to do on systems that don't support cuda? 

I tested it on my MacBook -- which doesn't support cuda -- and the test worked fine.

Reply:
...but it's failing weirdly on Windows :-/